services/src/main/java/org/keycloak/broker/saml/mappers/UsernameTemplateMapper.java

```diff
@@ -92,10 +93,15 @@
 
         TRANSFORMERS.put("uppercase", String::toUpperCase);
         TRANSFORMERS.put("lowercase", String::toLowerCase);
+        TRANSFORMERS.put("localpart", UsernameTemplateMapper::getEmailLocalPart);
     }
 
     public static final String PROVIDER_ID = "saml-username-idp-mapper";
 
+    public static String getEmailLocalPart(String email) {
+      return email.substring(0, email.lastIndexOf("@"));
```

Comment:
Apologies, I missed a potential for `IndexOutOfBoundsException` here in my preview review. `lastIndexOf` [can return negative value](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#lastIndexOf-int-) and this would [lead to the exception](https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#substring(int,%20int)) because -1 < 0

When at it, prefer the character form over the `String` one, ie.
```suggestion
      return email.substring(0, email.lastIndexOf('@'));
```

Reply:
Good point.

What's your opinion about the preferred return value in that case?

Empty string or original string?

Reply:
Good point. This should resolve it.
I made the transformer a no-op for such case and modified help text accordingly.